README.md
# Sunix ledstrip controller

A dart library for controlling the Sunix® RGB / RGBWWCW WiFi LED strip controller.

## Features

- Power the controller off and on
- Change the color of the led controller
  - For RGB or RGBWWCW (warm white and cold white) colors
- Receive the current status of the controller, including the power state and color values
- Send multiple requests in quick succession
  - Can be used to create a smooth transition from one color to another

## Installation

You can install the package by adding the following to your `pubspec.yaml`:

```yaml
dependencies:
  sunix_ledstrip_controller: ^1.0.2
```

## Usage

A simple usage example:

```dart
import 'package:sunix_ledstrip_controller/sunix_ledstrip_controller.dart';

main() async {
  // create a new controller object with a static ip
  LedController controller = LedController("192.168.178.50");

  // power the controller on
  if (await controller.powerOn()) {
    // methods return true when the request was successfully sent
    print("successfully powered on");
  }

  // change the controller color
  controller.updateColorRgb(192, 255, 238);

  // get the status of the controller
  StatusResponse status;
  await controller
      .requestStatus()
      .then((response) => status = response);

  print("powered on: ${status?.poweredOn}");
  print("red:   ${status?.red}");
  print("blue:  ${status?.green}");
  print("green: ${status?.blue}");
}
```

## Example

See the `example` directory for a basic example.

## Attributions

A big thanks goes to [@markusressel](https://github.com/markusressel) for his python library [sunix-ledstrip-controller-client](https://github.com/markusressel/sunix-ledstrip-controller-client).


CHANGELOG.md
## 1.0.2

- Remove logging dependency
- Cleanup dart docs

## 1.0.1

- Update readme


## 1.0.0

- Initial version


test/sunix_ledstrip_controller_test.dart
import 'package:sunix_ledstrip_controller/sunix_ledstrip_controller.dart';
import 'package:test/test.dart';

void main() {
  group('requests', () {
    test('update color rgb', () {
      final request = UpdateColorRequest.rgb(red: 128, blue: 128, green: 128);

      expect(request.red, equals(128));
      expect(request.green, equals(128));
      expect(request.blue, equals(128));

      expect(request.rgbwwSelection, UpdateColorRequest.selectionRgb);

      expect(request.data, equals([49, 128, 128, 128, 0, 0, 240, 15, 176]));
    });

    test('update color ww', () {
      final request = UpdateColorRequest.ww(warmWhite: 13, coldWhite: 37);

      expect(request.warmWhite, equals(13));
      expect(request.coldWhite, equals(37));

      expect(request.rgbwwSelection, UpdateColorRequest.selectionWw);

      expect(request.data, equals([49, 0, 0, 0, 13, 37, 15, 15, 129]));
    });

    test('update color rgbww', () {
      final request = UpdateColorRequest.rgbww(
        red: 0,
        green: 125,
        blue: 250,
        warmWhite: 13,
        coldWhite: 37,
      );

      expect(request.red, equals(0));
      expect(request.green, equals(125));
      expect(request.blue, equals(250));
      expect(request.warmWhite, equals(13));
      expect(request.coldWhite, equals(37));

      expect(request.rgbwwSelection, UpdateColorRequest.selectionRgbww);

      expect(request.data, equals([49, 0, 125, 250, 13, 37, 255, 15, 232]));
    });

    test('set power on', () {
      expect(SetPowerRequest.on().data, equals([113, 35, 15, 163]));
    });

    test('set power off', () {
      expect(SetPowerRequest.off().data, equals([113, 36, 15, 164]));
    });

    test('status', () {
      expect(StatusRequest().data, equals([129, 138, 139, 150]));
    });
  });

  test('calculate checksum', () {
    final request = UpdateColorRequest.rgb(red: 255, green: 0, blue: 255);

    expect(request.data.last, equals(46));
  });
}


lib/src/requests.dart
import 'package:sunix_ledstrip_controller/src/checksum.dart';

const localGateway = 0x0F;
const remoteGateway = 0xF0;

const powerStateOn = 0x23;
const powerStateOff = 0x24;

/// A request for the led controller.
abstract class LedRequest {
  /// Contains the data of a request package.
  List<int> data = [];

  /// Flag to signify the [LedController] to wait for a response after sending
  /// the request.
  bool waitForResponse = false;

  @override
  String toString() => data.map((value) => value.toRadixString(16)).toString();
}

/// Request for changing the color for the controller.
///
/// Throws an [ArgumentError] if any color is invalid.
class UpdateColorRequest extends LedRequest {
  static const selectionRgb = 0xF0;
  static const selectionWw = 0x0F;
  static const selectionRgbww = 0xFF;

  /// The identifier for the request.
  static const packetId = 0x31;

  /// The color values.
  int red = 0x00;
  int green = 0x00;
  int blue = 0x00;
  int warmWhite = 0x00;
  int coldWhite = 0x00;

  /// Specifies which colors to use.
  ///
  /// 0xF0: Only update [red], [green], [blue].
  /// 0x0F: Only update [warmWhite], [coldWhite].
  /// 0xFF: Update all.
  /// 0x00: Update none.
  int rgbwwSelection = selectionRgb;

  /// Specifies if the gateway is accessible locally or remotely.
  ///
  /// The remote value is only used by the official app.
  /// 0x0F: Local.
  /// 0xF0: Remote.
  int remoteOrLocal = localGateway;

  UpdateColorRequest.rgb({
    this.red,
    this.green,
    this.blue,
  }) : rgbwwSelection = selectionRgb {
    _addData();
  }

  UpdateColorRequest.ww({
    this.warmWhite,
    this.coldWhite,
  }) : rgbwwSelection = selectionWw {
    _addData();
  }

  UpdateColorRequest.rgbww({
    this.red,
    this.green,
    this.blue,
    this.warmWhite,
    this.coldWhite,
  }) : rgbwwSelection = selectionRgbww {
    _addData();
  }

  void _addData() {
    _validateColors();

    data
      ..add(packetId)
      ..add(red)
      ..add(green)
      ..add(blue)
      ..add(warmWhite)
      ..add(coldWhite)
      ..add(rgbwwSelection)
      ..add(remoteOrLocal)
      ..add(calculateChecksum(data));
  }

  /// Throws an [ArgumentError] if any color is invalid.
  void _validateColors() {
    final colors = <int>[
      red,
      green,
      blue,
      warmWhite,
      coldWhite,
    ];

    final invalidColors = colors
        .where((value) => value != null ? value < 0 || value > 255 : false);

    if (invalidColors.isNotEmpty) {
      throw ArgumentError('Invalid color range for $invalidColors.');
    }
  }

  @override
  String toString() => 'Update color request: ${super.toString()}';
}

/// Request for changing the power state of the controller.
class SetPowerRequest extends LedRequest {
  static const packetId = 0x71;

  int powerStatus = powerStateOff;
  int remoteOrLocal = localGateway;

  SetPowerRequest.on() {
    powerStatus = powerStateOn;
    _addData();
  }

  SetPowerRequest.off() {
    powerStatus = powerStateOff;
    _addData();
  }

  void _addData() {
    data
      ..add(packetId)
      ..add(powerStatus)
      ..add(remoteOrLocal)
      ..add(calculateChecksum(data));
  }

  @override
  String toString() => 'Set power request: ${super.toString()}';
}

/// Request for the current status of the controller.
class StatusRequest extends LedRequest {
  static const packetId = 0x81;

  int payload1 = 0x8A;
  int payload2 = 0x8B;

  StatusRequest() {
    waitForResponse = true;

    data
      ..add(packetId)
      ..add(payload1)
      ..add(payload2)
      ..add(calculateChecksum(data));
  }

  @override
  String toString() => 'Status request: ${super.toString()}';
}


lib/src/response.dart
import 'package:sunix_ledstrip_controller/src/checksum.dart';
import 'package:sunix_ledstrip_controller/src/requests.dart';

/// A response from the led controller.
///
/// Uses the overridden [_dataMap] to parse the [data] in the [parse] function.
/// If the checksum is valid [valid] will be true, false otherwise.
abstract class LedResponse {
  Map<String, int> _dataMap = {};

  bool get valid => evaluateChecksum(_dataMap.values);

  /// Parses the data and maps it in the [_dataMap].
  ///
  /// Throws a [ResponseParserException] if the response length does not match
  /// the expected length from the [_dataMap] or if the checksum is invalid.
  void parse(List<int> data) {
    if (_dataMap.length != data.length) {
      throw ResponseParserException('Response length mismatch.\n'
          'Expected length: ${_dataMap.length}.\n'
          'Received length: ${data.length}');
    }

    if (!evaluateChecksum(data)) {
      throw ResponseParserException('Invalid response checksum');
    }

    for (var i = 0; i < data.length; i++) {
      _dataMap[_dataMap.keys.elementAt(i)] = data[i];
    }
  }

  @override
  String toString() => _dataMap.toString();
}

/// The response that contains the status of the led controller.
///
/// Used together with [StatusRequest].
class StatusResponse extends LedResponse {
  int get red => _dataMap['red'];
  int get green => _dataMap['green'];
  int get blue => _dataMap['blue'];
  int get warmWhite => _dataMap['warmWhite'];
  int get coldWhite => _dataMap['coldWhite'];

  bool get poweredOn => _dataMap['powerStatus'] == powerStateOn;
  bool get poweredOff => _dataMap['powerStatus'] == powerStateOff;

  StatusResponse(List<int> data) {
    _dataMap = {
      'packetId': null,
      'deviceName': null,
      'powerStatus': null,
      'mode': null,
      'runStatus': null,
      'speed': null,
      'red': null,
      'green': null,
      'blue': null,
      'warmWhite': null,
      '_unknown1': null,
      'coldWhite': null,
      '_unknown2': null,
      'checkSum': null,
    };

    parse(data);
  }
}

class ResponseParserException implements Exception {
  final String message;

  ResponseParserException(this.message);

  @override
  String toString() => 'Exception: Unable to parse response\n$message';
}


lib/src/led_controller.dart
import 'dart:async';
import 'dart:io';

import 'package:sunix_ledstrip_controller/src/requests.dart';
import 'package:sunix_ledstrip_controller/src/response.dart';

const defaultPort = 5577;

// todo: discover controller

class LedController {
  /// The host ip address of the led controller.
  String host;

  /// The port of the led controller (5577 by default).
  int port;

  LedController(this.host, [this.port = defaultPort])
      : assert(host != null),
        assert(port != null);

  /// Sends a [LedRequest] to a socket with the given [host] and [port].
  ///
  /// Can throw a [SocketException].
  ///
  /// If the request does not require to wait for a response, it will return
  /// null after the request has been sent to the socket.
  /// Otherwise it will return the response as a List<int>.
  /// If no data has been received in the duration of
  /// [timeout], a [TimeoutException] will be thrown.
  Future<List<int>> sendRequest(
    LedRequest request, [
    Duration timeout = const Duration(seconds: 10),
  ]) async {
    final socket = await Socket.connect(host, port);
    socket.add(request.data);

    // add listener when waiting for response
    if (request.waitForResponse) {
      List<int> responseData;

      await for (var data in socket.timeout(timeout)) {
        if (data is List<int>) {
          socket.destroy();
          responseData = data;
        } else if (data is TimeoutException) {
          socket.destroy();
          throw data;
        }
        socket.destroy();
      }

      return responseData;
    } else {
      socket.destroy();
      return null;
    }
  }

  /// Sends a list of [LedRequest]s in an [duration] to a socket with the given
  /// [host] and [port].
  ///
  /// All requests will be send in sequence with a delay, so that all requests
  /// will be send in the given [duration].
  ///
  /// Can throw a [SocketException].
  Future<void> sendRequests(
    List<LedRequest> requests,
    Duration duration,
  ) async {
    final interval = Duration(
      microseconds: duration.inMicroseconds ~/ requests.length,
    );

    final socket = await Socket.connect(host, port);

    for (final request in requests) {
      socket.add(request.data);
      sleep(interval);
    }

    socket.destroy();
  }

  /// Changes the rgb color for the led controller.
  ///
  /// Returns `true` if the request was successfully sent to the controller,
  /// `false` otherwise.
  Future<bool> updateColorRgb(int r, int g, int b) async {
    try {
      await sendRequest(UpdateColorRequest.rgb(red: r, green: g, blue: b));
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Changes the rgbww color for the led controller.
  ///
  /// Returns `true` if the request was successfully sent to the controller,
  /// `false` otherwise.
  Future<bool> updateColorRgbww(int r, int g, int b, int ww, int cw) async {
    try {
      await sendRequest(UpdateColorRequest.rgbww(
        red: r,
        green: g,
        blue: b,
        warmWhite: ww,
        coldWhite: cw,
      ));
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Changes the ww color for the led controller.
  ///
  /// Returns `true` if the request was successfully sent to the controller,
  /// `false` otherwise.
  Future<bool> updateColorWw(int ww, int cw) async {
    try {
      await sendRequest(UpdateColorRequest.ww(warmWhite: ww, coldWhite: cw));
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Powers the led controller on.
  ///
  /// Returns `true` if the request was successfully sent to the controller,
  /// `false` otherwise.
  Future<bool> powerOn() async {
    try {
      await sendRequest(SetPowerRequest.on());
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Powers the led controller off.
  ///
  /// Returns `true` if the request was successfully sent to the controller,
  /// `false` otherwise.
  Future<bool> powerOff() async {
    try {
      await sendRequest(SetPowerRequest.off());
      return true;
    } catch (e) {
      return false;
    }
  }

  /// Requests the status of the led controller.
  ///
  /// Returns a [StatusResponse] after receiving it or `null` if there was a
  /// problem sending the request or if the response is not received in time.
  Future<StatusResponse> requestStatus() async {
    try {
      final data = await sendRequest(StatusRequest());
      return StatusResponse(data);
    } catch (e) {
      return null;
    }
  }
}


example/sunix_ledstrip_controller_example.dart
import 'dart:convert';
import 'dart:io';

void main() async {
  const discoveryPort = 48899;
  const discoveryMessage = 'HF-A11ASSISTHREAD';

  final udpSocket = await RawDatagramSocket.bind(
    InternetAddress.anyIPv4,
    discoveryPort,
  );

  udpSocket.broadcastEnabled = true;

  udpSocket.listen((data) {
    final dg = udpSocket.receive();
    if (dg != null) {
      print('data received: ${dg.data}');
      print(dg.address);
      print(dg.port);
      print(utf8.decode(dg.data));
    } else {
      print('data is null');
    }
  });

  udpSocket.send(
    utf8.encode(discoveryMessage),
    InternetAddress('255.255.255.255'),
    discoveryPort,
  );
}

// void main() async {
//   print('start example');

//   // create a new controller object with a static ip
//   final controller = LedController('192.168.178.50');

//   // power the controller on
//   if (await controller.powerOn()) {
//     // methods return true when the request was successfully sent
//     print('successfully powered on');
//   }

//   // change the controller color
//   await controller.updateColorRgb(192, 255, 238);

//   // also change the warm and cold white (if the controller supports it)
//   await controller.updateColorRgbww(192, 255, 238, 128, 128);

//   // get the status of the controller
//   StatusResponse status;

//   await controller
//       .requestStatus()
//       .then((response) => status = response ?? status);

//   print('powered on: ${status?.poweredOn}');
//   print('red:   ${status?.red}');
//   print('blue:  ${status?.green}');
//   print('green: ${status?.blue}');

//   // send a list of requests
//   // can be used to update the color in quick succession
//   final requests = <LedRequest>[
//     UpdateColorRequest.rgb(red: 0, green: 0, blue: 0),
//     UpdateColorRequest.rgb(red: 60, green: 0, blue: 0),
//     UpdateColorRequest.rgb(red: 120, green: 0, blue: 0),
//     UpdateColorRequest.rgb(red: 180, green: 0, blue: 0),
//     UpdateColorRequest.rgb(red: 240, green: 0, blue: 0),
//   ];

//   // when using sendRequest or sendRequests an exception can be thrown and
//   // should be caught
//   try {
//     await controller.sendRequests(requests, const Duration(seconds: 10));
//   } on SocketException {
//     print('exception while changing color');
//   }
// }
